### ISSUE TITLE
Library uses structs rather than interfaces

### ISSUE THREAD

**Issue by cholick** [orchestrate-io/gorc2#14]
Types like `Client`, `Collection`, `Iterator`, and others are all structs in the implementation of the library. This makes unit testing the collaboration with the Orchestrate Go library difficult.

For example, `Client.Collection()` returns a `*gorc2.Collection`.  It would be much easier to build unit tests around code collaborating with gorc2 if it returned an interface of type `gorc2.Collection`, so that we could supply our own mock implementation that satisfies that interface.  To actually mock `gorc2.Collection`, we've had to create an interface and a concrete type that delegates to gorc2.

**Example:**
```
type OioCollection interface {
	AddEvent(key, typ string, value interface{}) (*gorc2.Event, error)
	ListEvents(key, typ string, opts *gorc2.ListEventsQuery) *gorc2.Iterator

	Create(key string, value interface{}) (item *gorc2.Item, err error)
	Update(key string, value interface{}) (*gorc2.Item, error)
	Get(key string, value interface{}) (*gorc2.Item, error)
}

type OioClient interface {
	Collection(name string) OioCollection
}

type oioClient struct {
	gorcClient *gorc2.Client
}

func (client *oioClient) Collection(name string) OioCollection {
	return client.gorcClient.Collection(name)
}
```

*In the above code, we've abbreviated the interface wrapping the `Collection` struct.*

**Comment by liquidgecka** [comment]
The issue that pops up with interfaces is that they are a contract that basically requires zero extensions in order to maintain compatibility. I debated going the interface route when I initially wrote the client but decided against it because it doesn't fit real well with go's model.

For example, if I exposed Collection as an interface and you implemented it for your tests, then I couldn't easily add new features via a new function. Doing so would instantly break your existing implementation and make your tests fail to compile. And since we update the client quite frequently it would get super annoying to have people randomly broken every time they update (and therefor they would refuse to update which is bad for everybody involved).

Now, that being said, it is completely possible to manipulate the gorc2 client using the httpclient that is fully exposed. I use this is several instances in the testing for gorc2 directly. Its not as readable as an example as I would like, but basically I just created a structure that takes functions that get run either before or after the actual http call. Each test then sets these function to do something during the request cycle.

The class is here: https://github.com/orchestrate-io/gorc2/blob/master/util_test.go#L523
And an example of using it is here: https://github.com/orchestrate-io/gorc2/blob/master/client_test.go#L107

This could easily be replaced with a generic http RoundTripper replacement that can fully move the http call. Its more annoying than a mock class, but it does get to the same rough end point.

**Comment by cholick** [comment]
A disadvantage to this is that we have to construct a proper json response for the RoundTripper's body. This requires diving into and understanding the API to a level that the library would abstract away.

**Comment by liquidgecka** [comment]
Absolutely.. I was tempted a while ago to write a helper testing package for gorc2 (gorc2.testhelper or some such) that abstracts a ton of the normal gorc2 testing headaches. That is basically how dvr came to be in the first place..

In order to make a true interface based situation work it would basically end up needing to be a complete retooling since Item/Event both have significant work elements to them, so they would need an interface as well, which in turn changes a bunch of underlying stuff. =/

Let me noodle on thi a bit to see if I can come up with a clean solution.

**Comment by cholick** [comment]
Sounds good, thanks.
